### ISSUE TITLE
ts-proto expects Reader.int64() to always return Long, but protobufjs may return number

### ISSUE THREAD

**Issue by timostamm** [stephenh/ts-proto#78]
When protobufjs is used with a bundler, its Reader.int64() returns an unsafe `number` instead of `Long`. 

Can we add a "Gotcha: Module bundlers and long.js" section to the readme?

The following workaround forces protobufjs to use long.js:

```ts
import * as protobuf from 'protobufjs/minimal';
import * as Long from 'long';
protobuf.util.Long = Long;
protobuf.configure();
```
It should be included in the app´s bootstrap code, `main.ts` for angular.

This is required for every `forceLong` option. `forceLong=string` may not throw errors, but it will probably silently overflow the value.

**Comment by stephenh** [comment]
Hm. Great find/point @timostamm . Fwiw ts-proto working in the browser is almost a complete accident given I never focused on it initially. But I love that people are using it there.

This is a large ask, but could you make a PR that adds an `./integration-test` that somehow replicates the issue? I know right now all of the integration-tests run their unit tests in jest on node, but that's exactly why this sort of stuff isn't caught.

I'm technically somewhat-capable in the frontend ecosystem, but most/all of the react tests that I use on a day-to-day basis still run in node with jsdom, and I don't think I've setup a "no really bundle everything to `.js` and run the tests against that".

Is that something you could setup? And then reproduce this?

I'm also wondering if the ^ incantation you have could be something that `ts-proto` just does on initialization? Like when the module is imported, a) detect if this is the browser, b) "know" what `forceLong` option is being used, and c) do this `protobuf.util.Long = Long` automatically?

If you think that's a good idea, that's what I like to have that `integration-test/longs-in-browser` test to cover/make sure it works.

Thanks!

**Comment by stephenh** [comment]
@timostamm so, would it be useful to do something like:

* Add a global variable like `global.tsProtoHasInitializedProtobuf`
* In any file that ts-proto generates, if `forceLong` is on, include that snippet you have
  * The snippet will only be run if `global.tsProtoHasInitializedProtobuf` is unset, and then will set it, so that we only do it once even if you have multiple files generated by ts-proto
* (Optional) Add a `configureProtobuf` config flag which defaults `true` but can be set `false` (if for some reason projects don't want this "we'll just call configure for you" behavior)

If that makes sense to you, do want to work on a PR that does that?

**Comment by timostamm** [comment]
A `global.tsProtoHasInitializedProtobuf` might be unnecessary? The `if` body below runs only once:

```ts
if (protobuf.util.Long !== Long) {
  protobuf.util.Long = Long;
  protobuf.configure();
}
```

A `configureProtobuf` flag makes sense. 

I am not sure if it is sufficient to just render the snippet. This side effect might be stripped by a tree shaker... 
A simple test with a console.log statement is left intact in my first test, but I am not sure if I am triggering tree shaking correctly or if other shakers do this differently. 

If anybody has more experience on this topic, please chime in. 

I will do some tests when the opportunity comes up, but right now, time is tight.

**Comment by stephenh** [comment]
Oh right, `protobuf.util.Long !== Long`  as a guard to make sure we only do it once makes sense.

> This side effect might be stripped by a tree shaker...

Hm, I'm generally skeptical that a tree shaker would remove module-level initialization code from a module that is actively imported by the project code. That would seem like a bad thing for it to do.

**Comment by timostamm** [comment]
> Hm, I'm generally skeptical that a tree shaker would remove module-level initialization code from a module that is actively imported by the project code. That would seem like a bad thing for it to do.

After investigating this a little bit I agree. PR is on the way.

**Comment by Raiondesu** [comment]
If anyone stumbled upon this and is still waiting for the fix, you can install the patched version via
```
npm i Raiondesu/ts-proto#build
```

**Comment by stephenh** [comment]
@timostamm @Raiondesu does https://github.com/stephenh/ts-proto/pull/125 do what you had in mind?

**Comment by stephenh** [comment]
Oh shoot, @timostamm , I forgot about your https://github.com/stephenh/ts-proto/pull/112 ... sorry about that!

**Comment by timostamm** [comment]
> Oh shoot, @timostamm , I forgot about your #112 ... sorry about that!

No worries, @stephenh. The ball was in my hands and I did not respond in time. Updating my PR right now, getting back to you in #112.

**Comment by stephenh** [comment]
This should be fixed in v1.33.0.
